apollo-router/src/plugins/record_replay/record.rs

```diff
@@ -0,0 +1,350 @@
+use crate::plugin::Plugin;
+use crate::plugin::PluginInit;
+use crate::services::execution;
+use crate::services::router;
+use crate::services::subgraph;
+use crate::services::supergraph;
+use crate::spec::query::Query;
+use crate::spec::Schema;
+use crate::Configuration;
+
+const RECORD_HEADER: &str = "x-apollo-router-record";
+const RECORD: &str = "record";
+const CLIENT_REQUEST: &str = "client_request";
+const CLIENT_RESPONSE: &str = "client_response";
+const QUERY_PLAN: &str = "query_plan";
+const SUBGRAPHS: &str = "subgraphs";
+
+/// Request recording configuration.
+#[derive(Debug, Clone, serde::Deserialize, schemars::JsonSchema)]
+#[serde(deny_unknown_fields)]
+struct RecordConfig {
+    /// The recording plugin is disabled by default.
+    enabled: bool,
+    /// The path to the directory where recordings will be stored. Defaults to
+    /// the current working directory.
+    #[serde(default = "default_storage_path")]
+    storage_path: String,
+}
+
+fn default_storage_path() -> String {
+    std::env::current_dir()
+        .expect("failed to get current directory")
+        .to_str()
+        .expect("failed to convert current directory to string")
+        .to_string()
+}
+
+#[derive(Debug)]
+struct Record {
+    enabled: bool,
+    supergraph_sdl: Arc<String>,
+    storage_path: Arc<Path>,
+    schema: Arc<Schema>,
+}
+
+register_plugin!("experimental", "record", Record);
+
+#[async_trait::async_trait]
+impl Plugin for Record {
+    type Config = RecordConfig;
+
+    async fn new(init: PluginInit<Self::Config>) -> Result<Self, BoxError> {
+        let storage_path = PathBuf::from(init.config.storage_path);
+
+        write_file(
+            storage_path.clone().into(),
+            "README.md",
+            include_str!("recording-readme.md").as_bytes(),
+        )
+        .await?;
+
+        let plugin = Self {
+            enabled: init.config.enabled,
+            supergraph_sdl: init.supergraph_sdl.clone(),
+            storage_path: storage_path.into(),
+            schema: Arc::new(Schema::parse(
+                init.supergraph_sdl.clone().as_str(),
+                &Configuration::default(),
+            )?),
+        };
+
+        Ok(plugin)
+    }
+
+    fn router_service(&self, service: router::BoxService) -> router::BoxService {
+        if !self.enabled {
+            return service;
+        }
+
+        let dir = self.storage_path.clone();
+        let supergraph_sdl = self.supergraph_sdl.clone();
+
+        ServiceBuilder::new()
+            .map_future(move |future| {
+                let dir = dir.clone();
+                let supergraph_sdl = supergraph_sdl.clone();
+
+                async move {
+                    let res: router::Response = future.await?;
+                    let (parts, stream) = res.response.into_parts();
+
+                    let headers = parts.headers.clone();
+                    let context = res.context.clone();
+
+                    let stream = stream.after_complete(move || {
```

Comment:
the recording task should happen in a `spawn_blocking`, since serializing and writing are synchronous and could block